// @aerisweather/javascript-sdk/dist/network/maps/MapResult.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _MapMetadata = _interopRequireDefault(require("./MapMetadata"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * An `MapResult` object contains response information about an AMP request.
 */
class MapResult {
  /**
   * Initializes a result instance with the necessary response information.
   *
   * @param response
   * @param image
   * @param error
   * @param params
   */
  constructor(response, image, error, params = null) {
    this.response = response;
    this.image = image;
    this.error = error;
    this.params = params;

    if (response && response.headers) {
      this.metadata = new _MapMetadata.default(response.headers);
    }
  }
  /**
   * Returns the headers returned by the response, if any.
   */


  headers() {
    return this.response ? this.response.headers : undefined;
  }

}

var _default = MapResult;
exports.default = _default;
module.exports = exports.default;
