// @aerisweather/javascript-sdk/dist/maps/LayerTimes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Http = _interopRequireDefault(require("../network/Http"));

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class LayerTimes {
  constructor(account) {
    this.account = account;
  }

  times(layer, limit = 20) {
    return new Promise((resolve, reject) => {
      const url = `https://maps.aerisapi.com/${this.account.id}_${this.account.secret}/${layer}.json?limit=${limit}`;

      _Http.default.request(url).then(res => {
        let result = [];

        if (res.data) {
          const data = res.data;

          if ((0, _utils.isset)(data.files)) {
            result = data.files.map(time => ({
              time: new Date(time.timestamp * 1000),
              min: new Date(time.minTimestamp * 1000),
              max: new Date(time.maxTimestamp * 1000),
              run: new Date(time.runTime * 1000)
            }));
          }
        }

        resolve(result);
      }).catch(error => {
        reject(error);
      });
    });
  }

  timesInRange(layer, from, to, limit = 20) {
    return this.times(layer, limit).then(times => times.filter(valid => valid.time.getTime() >= from.getTime() && valid.time.getTime() <= to.getTime()));
  }

}

var _default = LayerTimes;
exports.default = _default;
module.exports = exports.default;
