// @aerisweather/javascript-sdk/dist/maps/interactive/strategies/openlayers/Popup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _DOM = _interopRequireDefault(require("../../../../display/DOM"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Popup {
  constructor(coordinate) {
    if (!document) return;
    const target = (0, _DOM.default)('<div class="aeris__openlayers__popup"></div>');
    this._container = target; // this._container.css({
    // 	background: 'rgba(0, 0, 0, 0.9)',
    // 	borderRadius: '6px',
    // 	bottom: '100%',
    // 	color: '#fff',
    // 	display: 'block',
    // 	fontSize: '11px',
    // 	fontWeight: 'normal',
    // 	left: '50%',
    // 	lineHeight: '14px',
    // 	marginBottom: '15px',
    // 	marginLeft: '-100px',
    // 	opacity: 0,
    // 	padding: '8px 10px',
    // 	pointerEvents: 'none',
    // 	position: 'absolute',
    // 	transition: 'opacity .2s ease-out',
    // 	width: '200px',
    // 	zIndex: 10000,
    // 	'-webkit-font-smoothing': 'antialiased'
    // });

    this._overlay = new ol.Overlay({
      insertFirst: true,
      element: target[0],
      autoPan: true,
      autoPanAnimation: {
        duration: 250,
        source: null
      }
    });

    if (coordinate) {
      this.setCoordinate(coordinate);
    }
  }

  addTo(map) {
    this._map = map;
    map.addOverlay(this._overlay); // adjust left margin now that we have a valid width for the content

    const w = this._container.width();

    if (w > 0) {
      this._container.css({
        marginLeft: `-${Math.round(w / 2)}px`
      });
    }
  }

  remove() {
    if (this._map) {
      this._map.removeOverlay(this._overlay);
    }
  }

  setCoordinate(coord) {
    const [x, y] = ol.proj.fromLonLat([coord.lon, coord.lat]);

    this._overlay.setPosition([Math.round(x), Math.round(y)]);
  }

  setHTML(html) {
    this._container.html(html);
  }

  setDOMContent(node) {
    this._container.empty().append(node);
  }

  static injectStyles() {
    const css = `
			.aeris__openlayers__popup {
				background: #fff;
				border-radius: 6px;
				bottom: 100%;
				color: #222;
				display: block;
				filter: drop-shadow(0 1px 4px rgba(0,0,0,0.2));
				font-family: 'Helvetica','Arial',sans-serif;
				font-size: 13px;
				font-weight: normal;
				left: 50%;
				line-height: 14px;
				margin-bottom: 15px;
				margin-left: -100px;
				min-width: 120px;
				padding: 8px 10px;
				pointer-events: none;
				position: absolute;
				transition: opacity .2s ease-out;
				z-index: 10000;
				-webkit-font-smoothing: antialiased
			}

			.aeris__openlayers__popup:after {
				border-left: solid transparent 10px;
				border-right: solid transparent 10px;
				border-top: solid #fff 10px;
				bottom: -10px;
				content: " ";
				height: 0;
				left: 50%;
				margin-left: -9px;
				position: absolute;
				width: 0
			}
		`;
    (0, _DOM.default)('head').append(`<style type="text/css">${css}</style>`);
  }

}

var _default = Popup;
exports.default = _default;
module.exports = exports.default;
