// @aerisweather/javascript-sdk/dist/maps/interactive/strategies/google/layers/PolygonLayer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Layer = _interopRequireDefault(require("./Layer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class GooglePolygonLayer extends _Layer.default {
  constructor(layer, polygons, opts = {}, id = undefined) {
    super(layer, opts, id);
    this._polygons = polygons;
  }

  setOpacity(value) {
    super.setOpacity(value);

    if (this.renderable) {
      this.renderable.forEach(polygon => polygon.setOptions(Object.assign(Object.assign({}, this.opts), {
        fillOpacity: this.opts.fillOpacity * value,
        strokeOpacity: this.opts.strokeOpacity * value
      })));
    }
  }

  createRenderable() {
    const groups = [];
    const opts = this.opts; // need to combine all polygon coordinate paths into a single grouped array for the multi
    // polygon layer

    this._polygons.forEach(shape => {
      const outer = shape.outer.map(coord => ({
        lat: coord.lat,
        lng: coord.lon
      }));
      const inner = shape.inner.map(polygon => polygon.map(coord => ({
        lat: coord.lat,
        lng: coord.lon
      })));
      opts.paths = [outer].concat(inner);
      const p = new google.maps.Polygon(Object.assign(Object.assign({}, opts), {
        fillOpacity: this.opts.fillOpacity * this.opacity,
        opacity: this.opts.strokeOpacity * this.opacity
      }));
      p.addListener('click', () => this.trigger('click', {
        polygon: this
      }));
      p.addListener('mouseover', () => this.trigger('mouseover', {
        polygon: this
      }));
      p.addListener('mouseout', () => this.trigger('mouseout', {
        polygon: this
      }));
      groups.push(p);
    });

    return groups;
  }

  addToMap(map) {
    if (this.renderable) {
      this.renderable.forEach(polygon => polygon.setMap(map));
    }
  }

  removeFromMap(map) {
    if (this.renderable) {
      this.renderable.forEach(polygon => polygon.setMap(null));
    }
  }

}

var _default = GooglePolygonLayer;
exports.default = _default;
module.exports = exports.default;
