// @aerisweather/javascript-sdk/dist/account/Account.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Permissions = _interopRequireDefault(require("./Permissions"));

var _ApiRequest = _interopRequireDefault(require("../network/api/ApiRequest"));

var _MapRequest = _interopRequireDefault(require("../network/maps/MapRequest"));

var _globals = require("../globals");

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Account {
  constructor(id, secret) {
    this.id = id;
    this.secret = secret;
    this.servers = _globals.servers;
    this._permissions = new _Permissions.default(id, secret);
  }

  credentials() {
    return {
      id: this.id,
      secret: this.secret
    };
  }

  permissions() {
    return this._permissions;
  }

  canAccess(endpoint) {
    return this.permissions().canAccess(endpoint);
  }
  /**
   * Creates and returns a new ApiRequest configured for the account.
   *
   * @returns {ApiRequest}
   * @memberof Account
   */


  api(config) {
    return new _ApiRequest.default((0, _utils.extend)({
      server: this.servers.api,
      client: {
        id: this.id,
        secret: this.secret
      }
    }, config));
  }
  /**
   * Creates and returns a new MapRequest configured for the account.
   *
   * @returns {MapRequest}
   * @memberof Account
   */


  map(config) {
    return new _MapRequest.default((0, _utils.extend)({
      server: this.servers.maps,
      client: {
        id: this.id,
        secret: this.secret
      }
    }, config));
  }

}

var _default = Account;
exports.default = _default;
module.exports = exports.default;
