// @aerisweather/javascript-sdk/dist/modules/aviation/AviationGroup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _ModuleGroup = _interopRequireDefault(require("../ModuleGroup"));

var _flightrules = require("./flightrules");

var _sigmets = require("./sigmets");

var _stationplots = require("./stationplots");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var __awaiter = void 0 && (void 0).__awaiter || function (thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }

  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }

    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }

    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }

    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};

/**
 * The Aviation module contains a series of modules that are specific to aviation usage.
 *
 * Includes:
 * - flight rules
 * - SIGMETs
 * - station plots
 *
 * @class AviationModule
 * @extends {ModuleGroup}
 */
class AviationGroup extends _ModuleGroup.default {
  get id() {
    return 'aviation';
  }

  controls() {
    return {
      title: 'Aviation',
      buttons: this.modules ? this.modules.map(m => m.controls()) : []
    };
  }

  load() {
    return __awaiter(this, void 0, void 0, function* () {
      const flightRules = yield (0, _flightrules.loader)();
      const stationPlots = yield (0, _sigmets.loader)();
      const sigmets = yield (0, _stationplots.loader)();
      return new Promise((resolve, reject) => {
        this._modules = [flightRules, stationPlots, sigmets].map(Module => new Module.default()); // eslint-disable-line new-cap

        resolve(this._modules);
      });
    });
  }

}

var _default = AviationGroup;
exports.default = _default;
module.exports = exports.default;
